org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/resource/BatchLinkableResourceStorageLoadable.java

```diff
@@ -0,0 +1,112 @@
+import org.eclipse.xtext.xbase.jvmmodel.JvmModelAssociator;
+
+import com.google.common.collect.FluentIterable;
+
+public class BatchLinkableResourceStorageLoadable extends ResourceStorageLoadable {
+
+	public BatchLinkableResourceStorageLoadable(InputStream in, boolean storeNodeModel) {
+		super(in, storeNodeModel);
+	}
+
+	@Override
+	protected void loadEntries(StorageAwareResource resource, ZipInputStream zipIn) throws IOException {
+		super.loadEntries(resource, zipIn);
+		if (resource instanceof BatchLinkableResource) {
+			try {
+				readAssociationsAdapter((BatchLinkableResource) resource, zipIn);
+			} catch(ClassNotFoundException e) {
+				throw new IOException(e);
+			}
+		}
+	}
+
+	@Override
+	protected void handleLoadEObject(InternalEObject loaded, BinaryResourceImpl.EObjectInputStream input)
+			throws IOException {
+		super.handleLoadEObject(loaded, input);
+		if (input.readBoolean()) {
+			DocumentationAdapter documentationAdapter = new DocumentationAdapter(input.readString());
+			loaded.eAdapters().add(documentationAdapter);
+		}
+		if (input.readBoolean()) {
+			JvmIdentifiableMetaData metaData = new JvmIdentifiableMetaData();
+			metaData.setSynthetic(input.readBoolean());
+			loaded.eAdapters().add(metaData);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	protected void readAssociationsAdapter(BatchLinkableResource resource, ZipInputStream stream) throws IOException, ClassNotFoundException {
+		JvmModelAssociator.Adapter adapter = FluentIterable.from(resource.eAdapters())
+				.filter(JvmModelAssociator.Adapter.class)
+				.first()
+				.or(()->{
+					JvmModelAssociator.Adapter newAdapter = new JvmModelAssociator.Adapter();
+					resource.eAdapters().add(newAdapter);
+					return newAdapter;
+				});
+		
+		stream.getNextEntry();
+		
+		try (ObjectInputStream objIn = new ObjectInputStream(new NonClosingFilterInputStream(new BufferedInputStream(stream)))) {
+			Map<String, String> logicalMap = (Map<String, String>) objIn.readObject();
+			logicalMap.forEach((key, value)->{
+				adapter.logicalContainerMap.put(resource.getEObject(key), (JvmIdentifiableElement)resource.getEObject(value));
+			});
+			Map<String, Set<String>> sourceToTargetMap = (Map<String, Set<String>>) objIn.readObject();
+			sourceToTargetMap.forEach((source, targets)->{
+				adapter.sourceToTargetMap.put(resource.getEObject(source), fragmentsToObjects(resource, targets));
+			});
+			Map<String, Set<String>> targetToSourceMap = (Map<String, Set<String>>) objIn.readObject();
+			targetToSourceMap.forEach((target, sources)->{
+				adapter.targetToSourceMap.put(resource.getEObject(target), fragmentsToObjects(resource, sources)); 
+			});
+		}
+	}
+
+	private Set<EObject> fragmentsToObjects(BatchLinkableResource resource, Set<String> fragments) {
+		return FluentIterable.from(fragments)
+		.transform(resource::getEObject)
+		.copyInto(new LinkedHashSet<>());
+	}
+	
+	private static class NonClosingFilterInputStream extends FilterInputStream {
```

Comment:
buffering?

Reply:
Extend buffering instead of decorate buffering? Is that what you mean?

Reply:
the old used a bufferedInputstream

Reply:
`new ObjectInputStream(new NonClosingFilterInputStream(new BufferedInputStream(stream))) `

It still does

Reply:
looks like i missglanced this one